### ISSUE TITLE
Make L045 (Query defines a CTE but does not use it) case insensitive

### ISSUE THREAD

**Issue by jamiet-msm** [sqlfluff/sqlfluff#2675]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Running sqlfluff on a file returned 

> L: 186 | P:   1 | L045 | Query defines CTE "CRMEmailPimpLstWide" but does not use it.

The CTE _is_ used so I consider this to be a false positive.

note that this is in version 0.10.1. The error did not get raised in version 0.9.4 which is what we were using previously.

Possibly the same as https://github.com/sqlfluff/sqlfluff/issues/1999, but I'm not certain

### Expected Behaviour

No such error should be raised

### Observed Behaviour

> L: 186 | P:   1 | L045 | Query defines CTE "CRMEmailPimpLstWide" but does not use it.

### How to reproduce

Here's the file, notice that `CRMEmailPimpLstWide` _is_ used:

```sql
WITH vKeyToEKey AS ( --noqa: L045
    SELECT
        fe.Visitorkey,
        fe.EmailAddressKey,
        ROW_NUMBER() OVER (PARTITION BY fe.VisitorKey, dsrc.SourceType ORDER BY fe.EnquiryDateTime DESC) AS rnum
    FROM `${datalake_project_id}.edw_red.REDFctEnquiry` AS fe
    INNER JOIN `${project_containing_tables_required_for_braze}.edw.DimChannel` AS dchnl
        ON fe.ChannelKey = dchnl.ChannelKey
    INNER JOIN `${project_containing_tables_required_for_braze}.edw.DimSource` AS dsrc
        ON fe.SourceKey = dsrc.SourceKey
    WHERE UPPER(dchnl.BusinessGroupCode) = 'MS'
        AND UPPER(dsrc.SourceCode) NOT IN ('COIN', 'TIV', 'HQI', 'MQI', 'EQI')
        AND UPPER(dsrc.SourceType) != 'TEST'
        AND fe.EnquiryDateTime < TIMESTAMP_TRUNC(CURRENT_TIMESTAMP(), DAY)
        AND fe.EnquiryDateTime >= CAST(DATE_SUB(CURRENT_DATE(), INTERVAL 3 YEAR) AS TIMESTAMP)
),

CRMEmailVisitorKey2EmailKey
AS (
    SELECT
        VisitorKey,
        EmailAddressKey
    FROM vKeyToEKey
    WHERE rnum = 1
),

delta_changes AS (
    SELECT fe.EmailAddressKey
    FROM `${datalake_project_id}.edw_red.REDFctEnquiry` AS fe
    INNER JOIN `${project_containing_tables_required_for_braze}.edw.DimChannel` AS dchnl
        ON fe.ChannelKey = dchnl.ChannelKey
    INNER JOIN `${project_containing_tables_required_for_braze}.edw.DimSource` AS dsrc
        ON fe.SourceKey = dsrc.SourceKey
    WHERE fe.EnquiryDateTime >= CAST(DATE_SUB(CURRENT_DATE(), INTERVAL 3 YEAR) AS TIMESTAMP) AND fe.ExtractAuditKey > COALESCE(
        (
            SELECT HWMTimeStamp
            FROM `${audit_dataset_id}.${HighWaterMark_table_id}`
            WHERE Process = 'MSM_IN_MODEL.UserProfile_Events_WebApp'
            ORDER BY CompletedTime DESC
            LIMIT 1
        ), 0
        )
        AND (fe.IsSystemGenerated) != 'Yes'
        AND (dsrc.sourceType) != 'TEST'
        AND (dchnl.BusinessGroupCode) = 'MS'
    UNION DISTINCT
    SELECT v2e.EmailAddressKey
    FROM `${datalake_project_id}.edw_red.REDFctClick` AS click
    INNER JOIN CRMEmailVisitorKey2EmailKey AS v2e ON click.VisitorKey = v2e.visitorKey
    INNER JOIN `${project_containing_tables_required_for_braze}.edw.DimChannel` AS dchnl
        ON click.ChannelKey = dchnl.ChannelKey
    INNER JOIN `${project_containing_tables_required_for_braze}.edw.DimSource` AS dsrc
        ON click.SourceKey = dsrc.SourceKey
    WHERE click.ClickDateTime < TIMESTAMP_TRUNC(CURRENT_TIMESTAMP(), DAY) AND click.ExtractAuditKey > COALESCE((
        SELECT HWMTimeStamp
        FROM `${audit_dataset_id}.${HighWaterMark_table_id}`
        WHERE Process = 'MSM_IN_MODEL.UserProfile_Events_WebApp_Click'
        ORDER BY CompletedTime DESC
        LIMIT 1), 0)
        AND (dsrc.SourceCode) NOT IN ('COIN', 'TIV') AND (dsrc.SourceType) != 'TEST' AND (dchnl.BusinessGroupCode) = 'MS'
    UNION DISTINCT
    SELECT v2e.EmailAddressKey
    FROM `${project_containing_tables_required_for_braze}.edw.PageImpression` AS pimp
    INNER JOIN CRMEmailVisitorKey2EmailKey AS v2e
        ON pimp.VisitorKey = v2e.visitorKey
    INNER JOIN `${project_containing_tables_required_for_braze}.edw.DimChannel` AS dchnl
        ON pimp.ChannelKey = dchnl.ChannelKey
    INNER JOIN `${project_containing_tables_required_for_braze}.edw.DimSource` AS dsrc
        ON pimp.SourceKey = dsrc.SourceKey
    WHERE pimp.PimpDateTime < TIMESTAMP_TRUNC(CURRENT_TIMESTAMP(), DAY) AND pimp.ExtractAuditKey > COALESCE((
        SELECT HWMTimeStamp
        FROM `${audit_dataset_id}.${HighWaterMark_table_id}`
        WHERE Process = 'MSM_IN_MODEL.UserProfile_Events_WebApp_Pimp'
        ORDER BY CompletedTime DESC
        LIMIT 1), 0)
        AND (dsrc.SourceCode) NOT IN ('COIN', 'TIV') AND (dsrc.SourceType) != 'TEST' AND (dchnl.BusinessGroupCode) = 'MS'
    UNION DISTINCT
    SELECT tpf.EmailAddressKey
    FROM `${datalake_project_id}.edw_red.REDEmailDataMartTpf` AS tpf
    WHERE tpf.ExtractAuditKey > COALESCE((
        SELECT HWMTimeStamp
        FROM `${audit_dataset_id}.${HighWaterMark_table_id}`
        WHERE Process = 'MSM_IN_MODEL.UserProfile_Events_WebApp_Tpf'
        ORDER BY CompletedTime DESC
        LIMIT 1), 0)
    UNION DISTINCT
    SELECT uc.EmailAddressKey
    FROM `${datalake_project_id}.${Braze_Currents_UsersMessagesEmailOpen_dataset_id}.${Braze_Currents_UsersMessagesEmailOpen_table_id}` AS emailOpen
    INNER JOIN `${datalake_project_id}.edw_red.REDFctUserConsent` AS uc
        ON emailOpen.external_user_id = uc.UserCONsentId
            AND uc.UserCONsentUpdatedDateTime >= CAST(DATE_SUB(CURRENT_DATE(), INTERVAL 3 YEAR) AS TIMESTAMP)
            AND UPPER(uc.UserCONsentBrandName) = 'MONEYSUPERMARKET'
    WHERE emailOpen.time > COALESCE((
        SELECT HWMTimeStamp
        FROM `${audit_dataset_id}.${HighWaterMark_table_id}`
        WHERE Process = 'MSM_IN_MODEL.UserProfile_Events_WebApp_BrazeOpen'
        ORDER BY CompletedTime DESC
        LIMIT 1), 0)
    UNION DISTINCT
    SELECT uc.EmailAddressKey
    FROM `${datalake_project_id}.${Braze_Currents_UsersMessagesEmailClick_dataset_id}.${Braze_Currents_UsersMessagesEmailClick_table_id}` AS click
    INNER JOIN `${datalake_project_id}.edw_red.REDFctUserConsent` AS uc
        ON click.external_user_id = uc.UserCONsentId
            AND uc.UserCONsentUpdatedDateTime >= CAST(DATE_SUB(CURRENT_DATE(), INTERVAL 3 YEAR) AS TIMESTAMP)
            AND UPPER(uc.UserCONsentBrandName) = 'MONEYSUPERMARKET'
    WHERE click.time > COALESCE((
        SELECT HWMTimeStamp
        FROM `${audit_dataset_id}.${HighWaterMark_table_id}`
        WHERE Process = 'MSM_IN_MODEL.UserProfile_Events_WebApp_BrazeClick'
        ORDER BY CompletedTime DESC
        LIMIT 1), 0)
),

CRMEmailLstAppWebEnq AS (
    SELECT
        lenq.EmailAddressKey,
        MAX(CASE WHEN dsrc.SourceCode IN( 'MOB-MOREMONEY', 'MOB-PNX') THEN fe.EnquiryDateTime END) AS LstAppEnquiryDateTime,
        MAX(CASE WHEN dsrc.SourceCode NOT IN( 'MOB-MOREMONEY', 'MOB-PNX') THEN fe.EnquiryDateTime END) AS LstWebEnquiryDateTime,
        MAX(lenq.ExtractAuditKey) AS ExtractAuditKey
    FROM (
        SELECT
            LstEnq.EmailAddressKey,
            LstEnq.EnquiryKey,
            LstEnq.ExtractAuditKey
        FROM (
                SELECT
                    fe.EmailAddressKey,
                    fe.EnquiryKey,
                    fe.ExtractAuditKey,
                    ROW_NUMBER() OVER (PARTITION BY fe.EmailAddressKey, dsrc.SourceType ORDER BY fe.EnquiryDateTime DESC, fe.EnquiryKey DESC) AS rnum
                FROM delta_changes AS del
                INNER JOIN `${datalake_project_id}.edw_red.REDFctEnquiry` AS fe
                    ON del.EmailAddressKey = fe.EmailAddressKey --noqa
                INNER JOIN `${project_containing_tables_required_for_braze}.edw.DimChannel` AS dchnl
                    ON fe.ChannelKey = dchnl.ChannelKey --noqa
                INNER JOIN `${project_containing_tables_required_for_braze}.edw.DimSource` AS dsrc
                    ON fe.SourceKey = dsrc.SourceKey
                WHERE (dchnl.BusinessGroupCode) = 'MS'
                    AND (dsrc.SourceCode) NOT IN ('COIN', 'TIV', 'EQI', 'MQI', 'HQI')
                    AND (dsrc.SourceType) != 'TEST'
                    AND fe.EnquiryDateTime >= CAST(DATE_SUB(CURRENT_DATE(), INTERVAL 3 YEAR) AS TIMESTAMP)
                    AND fe.EnquiryDateTime < TIMESTAMP_TRUNC(CURRENT_TIMESTAMP(), DAY)
                    AND fe.EnquiryKey NOT IN (
                        SELECT COALESCE(eventkey, -1) AS eventkey
                        FROM `${project_containing_tables_required_for_braze}.edw.FctExclusion`
                        WHERE channelkey IN (23, 65, 57, 46, 66, 37, 68, 22, 92, 3)
                    )
        ) AS LstEnq
        WHERE LstEnq.rnum = 1
    ) AS lenq
    INNER JOIN `${datalake_project_id}.edw_red.REDFctEnquiry` AS fe
        ON lenq.EnquiryKey = fe.EnquiryKey
            AND fe.EnquiryDateTime < TIMESTAMP_TRUNC(CURRENT_TIMESTAMP(), DAY)
    INNER JOIN `${project_containing_tables_required_for_braze}.edw.DimSource` AS dsrc
        ON fe.SourceKey = dsrc.SourceKey
    GROUP BY lenq.EmailAddressKey
),

CRMEmailWebClickWide AS (
    SELECT
        v2e.EmailAddressKey,
        MAX(CASE WHEN dsrc.SourceCode IN( 'MOB-MOREMONEY', 'MOB-PNX') THEN click.ClickDateTime END) AS LstAppClickDateTime,
        MAX(CASE WHEN dsrc.SourceCode NOT IN( 'MOB-MOREMONEY', 'MOB-PNX') THEN click.ClickDateTime END) AS LstWebClickDateTime,
        MAX(click.ExtractAuditKey) AS LatestClickAuditKey
    FROM (
        SELECT
            VisitorKey,
            sourceKey,
            ChannelKey,
            ClickDateTime,
            ExtractAuditKey
        FROM `${datalake_project_id}.edw_red.REDFctClick` WHERE ClickDateTime < TIMESTAMP_TRUNC(CURRENT_TIMESTAMP(), DAY) ORDER BY VisitorKey
    ) AS click
    INNER JOIN CRMEmailVisitorKey2EmailKey AS v2e
        ON click.VisitorKey = v2e.visitorKey
    INNER JOIN delta_changes AS del
        ON v2e.EmailAddressKey = del.EmailAddressKey
    INNER JOIN `${project_containing_tables_required_for_braze}.edw.DimSource` AS dsrc
        ON click.sourceKey = dsrc.sourceKey
    INNER JOIN `${project_containing_tables_required_for_braze}.edw.DimChannel` AS dch
        ON click.ChannelKey = dch.ChannelKey
    WHERE (dsrc.SourceCode) NOT IN ('COIN', 'TIV') AND (dsrc.SourceType) != 'TEST' AND (dch.BusinessGroupCode) = 'MS'
    GROUP BY v2e.emailAddressKey
),

CRMEmailPimpLstWide AS (
    SELECT
        EmailAddressKey,
        MAX(CASE WHEN SourceCode IN( 'MOB-MOREMONEY', 'MOB-PNX') THEN lstPIMPDateTime END) AS LstAppPimpDateTime,
        MAX(CASE WHEN SourceCode NOT IN( 'MOB-MOREMONEY', 'MOB-PNX') THEN lstPIMPDateTime END) AS LstWebPimpDateTime,
        MAX(ExtractAuditKey) AS LatestPimpAuditKey
    FROM (
        SELECT
            v2e.EmailAddressKey,
            dsrc.SourceCode,
            MAX(pimp.PIMPDateTime) AS lstPIMPDateTime,
            MAX(pimp.ExtractAuditKey) AS ExtractAuditKey
        FROM `${project_containing_tables_required_for_braze}.edw.PageImpression` AS pimp
        INNER JOIN CRMEmailVisitorKey2EmailKey AS v2e
            ON pimp.VisitorKey = v2e.visitorKey
        INNER JOIN delta_changes AS del ON v2e.EmailAddressKey = del.EmailAddressKey
        INNER JOIN `${project_containing_tables_required_for_braze}.edw.DimSource` AS dsrc
            ON pimp.sourceKey = dsrc.sourceKey
        INNER JOIN `${project_containing_tables_required_for_braze}.edw.DimChannel` AS dch
            ON pimp.ChannelKey = dch.ChannelKey
        WHERE pimp.PIMPDateTime < TIMESTAMP_TRUNC(CURRENT_TIMESTAMP(), DAY)
            AND (dsrc.SourceCode) NOT IN ('COIN', 'TIV') AND (dsrc.SourceType) != 'TEST' AND (dch.BusinessGroupCode) = 'MS'
        GROUP BY v2e.emailAddressKey, dsrc.SourceCode
    )
    GROUP BY EmailAddressKey
),

CRMEmailBrazeActivity AS (
    SELECT
        external_id,
        EmailaddressKey,
        LstEmailActivityDate,
        LatestOpenTime,
        LatestClickTime
    FROM
        (
            SELECT
                external_id,
                EmailaddressKey,
                OpenTime AS LatestOpenTime,
                ClickTime AS LatestClickTime,
                TIMESTAMP_SECONDS(LstEmailActivityTime) AS LstEmailActivityDate,
                ROW_NUMBER() OVER (PARTITION BY EmailAddressKey ORDER BY LstEmailActivityTime DESC) AS rnum
            FROM (
                SELECT
                    uc.UserConsentId AS external_id,
                    uc.EmailaddressKey,
                    o.Time AS OpenTime,
                    c.Time AS ClickTime,
                    CASE
                        WHEN (o.Time IS NOT NULL AND c.Time IS NOT NULL) THEN GREATEST(o.Time, c.Time)
                        WHEN (o.Time IS NOT NULL AND c.Time IS NULL) THEN o.Time
                        WHEN (o.Time IS NULL AND c.Time IS NOT NULL) THEN c.Time
                    END AS LstEmailActivityTime
                FROM `${datalake_project_id}.edw_red.REDFctUserConsent` AS uc
                INNER JOIN delta_changes AS del
                    ON uc.EmailAddressKey = del.EmailAddressKey
                LEFT JOIN `${datalake_project_id}.${Braze_Currents_UsersMessagesEmailOpen_dataset_id}.${Braze_Currents_UsersMessagesEmailOpen_table_id}` AS o
                    ON uc.UserConsentId = o.external_user_id
                LEFT JOIN `${datalake_project_id}.${Braze_Currents_UsersMessagesEmailClick_dataset_id}.${Braze_Currents_UsersMessagesEmailClick_table_id}` AS c
                    ON uc.UserConsentId = c.external_user_id
                WHERE uc.UserConsentUpdatedDateTime >= CAST(DATE_SUB(CURRENT_DATE(), INTERVAL 3 YEAR) AS TIMESTAMP) AND UPPER(uc.UserConsentBrandName) = 'MONEYSUPERMARKET'
            )
        )
    WHERE rnum = 1
),

all_data AS (
    SELECT
        ceb.EmailaddressKey,
        uc.UserConsentId AS ExternalId,
        appwebenq.ExtractAuditKey,
        webclk.LatestClickAuditKey,
        pimp.LatestPimpAuditKey,
        tpf.ExtractAuditKey AS LatestTpfAuditKey,
        brazeactivity.LatestOpenTime,
        brazeactivity.LatestClickTime,
        LOWER(ceb.Emailaddress) AS Emailaddress,
        GREATEST(COALESCE(pimp.LstWebPimpDateTime, TIMESTAMP("1500-01-01 00:00:00")), COALESCE(webclk.LstWebClickDateTime, TIMESTAMP("1500-01-01 00:00:00")), COALESCE(appwebenq.LstWebEnquiryDateTime, TIMESTAMP("1500-01-01 00:00:00"))) AS LstWebActivityDateTime,
        GREATEST(COALESCE(pimp.LstAppPimpDateTime, TIMESTAMP("1500-01-01 00:00:00")), COALESCE(webclk.LstAppClickDateTime, TIMESTAMP("1500-01-01 00:00:00")), COALESCE(appwebenq.LstAppEnquiryDateTime, TIMESTAMP("1500-01-01 00:00:00"))) AS LstAppActivityDateTime,
        GREATEST(COALESCE(tpf.LstEmailActivityDate, TIMESTAMP("1500-01-01 00:00:00")), COALESCE(brazeactivity.LstEmailActivityDate, TIMESTAMP("1500-01-01 00:00:00"))) AS LstEmailActivityDate
    FROM delta_changes AS del
    INNER JOIN `${datalake_project_id}.edw_red.REDDimEmailAddress` AS ceb
        ON ceb.EmailAddressKey = del.EmailAddressKey
    LEFT JOIN CRMEmailPimpLstwide AS pimp
        ON pimp.EmailAddressKey = ceb.EmailAddressKey
    LEFT JOIN CRMEmailWebClickWide AS webclk
        ON webclk.EmailAddressKey = ceb.EmailAddressKey
    LEFT JOIN CRMEmailLstAppWebEnq AS appwebenq
        ON appwebenq.EmailAddressKey = ceb.EmailAddressKey
    LEFT JOIN CRMEmailBrazeActivity AS brazeactivity
        ON brazeactivity.EmailAddressKey = ceb.EmailAddressKey
    LEFT JOIN `${datalake_project_id}.edw_red.REDEmailDataMartTpf` AS tpf
        ON tpf.EmailAddressKey = ceb.EmailAddressKey
            AND tpf.LstEmailActivityDate >= CAST(DATE_SUB(CURRENT_DATE(), INTERVAL 3 YEAR) AS TIMESTAMP)
    INNER JOIN `${datalake_project_id}.edw_red.REDFctUserConsent` AS uc
        ON ceb.EmailAddressKey = uc.EmailAddressKey
            AND uc.UserConsentUpdatedDateTime >= CAST(DATE_SUB(CURRENT_DATE(), INTERVAL 3 YEAR) AS TIMESTAMP)
            AND UPPER(uc.UserCONsentBrandName) = 'MONEYSUPERMARKET'
)

SELECT
    EmailaddressKey,
    ExternalId,
    Emailaddress,
    CASE WHEN MAX(LstWebActivityDateTime) != TIMESTAMP("1500-01-01 00:00:00") THEN MAX(LstWebActivityDateTime) END AS LstWebActivityDateTime,
    CASE WHEN MAX(LstAppActivityDateTime) != TIMESTAMP("1500-01-01 00:00:00") THEN MAX(LstAppActivityDateTime) END AS LstAppActivityDateTime,
    CASE WHEN MAX(LstEmailActivityDate) != TIMESTAMP("1500-01-01 00:00:00") THEN MAX(LstEmailActivityDate) END AS LstEmailActivityDate,
    MAX(ExtractAuditKey) AS ExtractAuditKey,
    MAX(LatestClickAuditKey) AS LatestClickAuditKey,
    MAX(LatestPimpAuditKey) AS LatestPimpAuditKey,
    MAX(LatestTpfAuditKey) AS LatestTpfAuditKey,
    MAX(LatestOpenTime) AS LatestOpenTime,
    MAX(LatestClickTime) AS LatestClickTime
FROM all_data
WHERE LstWebActivityDateTime != TIMESTAMP("1500-01-01 00:00:00") OR LstAppActivityDateTime != TIMESTAMP("1500-01-01 00:00:00") OR LstEmailActivityDate != TIMESTAMP("1500-01-01 00:00:00")
GROUP BY EmailaddressKey, ExternalId, Emailaddress

```

running `sqlfluff lint --dialect bigquery --exclude-rules L016,L031,L014,L057 /path/to/file.sql`

returns : 
> L: 186 | P:   1 | L045 | Query defines CTE "CRMEmailPimpLstWide" but does not use
>                       | it.

Note that the templating used here is [terraform templating](https://www.terraform.io/language/functions/templatefile) however we've used this templating in many many files up to now with great success so I don't think that's the cause of the issue. 

### Dialect

bigquery

### Version

sqlfluff, version 0.10.1

### Configuration

n/a 
We use command-line options (provided above)

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tunetheweb** [comment]
It's `CRMEmailPimpLstWide` in the CTE but `CRMEmailPimpLstwide` (with small w) in the actual use.

Using the same case prevents the error.

Probably should make L045 case insensitive.

**Comment by jamiet-msm** [comment]
> It's `CRMEmailPimpLstWide` in the CTE but `CRMEmailPimpLstwide` (with small w) in the actual use.
> 
> Using the same case prevents the error.
> 
> Probably should make L045 case insensitive.

DOH. Good spot. Should have seen that, sorry.

**Comment by tunetheweb** [comment]
No worries, have renamed it and will keep it open as think we should make it case insensitive.
